iree/compiler/Dialect/Flow/IR/FlowOpUtils.cpp

```diff
@@ -14,131 +14,297 @@
+void eraseRegionResults(Region &region,
+                        ArrayRef<unsigned> excludedResultIndices) {
+  region.walk([&](IREE::Flow::ReturnOp terminator) {
+    llvm::SmallVector<Value, 4> newReturns;
+    for (auto it : llvm::enumerate(terminator.getOperands())) {
+      if (!llvm::count(excludedResultIndices, it.index())) {
+        newReturns.push_back(it.value());
+      }
+    }
+    terminator.getOperation()->setOperands(newReturns);
+  });
+}
+
+// Returns true if |constantOp| represents a (logically) small constant value.
+//
+// "Small" is relative and there's a risk that we'll bloat the closures by
+// duplicating a bunch of constants however what we are able to save by not
+// doing that usually wins. Think of the number of bytes used on instructions to
+// allocate/place/copy, setup function call arguments, compute the address,
+// dereference the memory, etc vs. a constant immediate value of 16 bytes -
+// afterall, there are single x64 instructions that approach 15 bytes :)
+//
+// This is also still at a fairly high level (flow dialect): once the closures
+// are expanded out in lower dialects things like CSE have a chance to once
+// again get at the constants and dedupe them if they survive.
+static bool isConstantSmall(ConstantOp constantOp) {
+  // We could tune this/take it as a configuration setting.
+  // The current value is chosen based on what is known to be reasonable to
+  // inline into command buffers way down in the HAL, which is not great but at
+  // least better than either allocating independent buffers for 4 byte
+  // constants or inlining megabytes.
+  static constexpr int kMaxInlinedConstantBytes = 256;
+
+  auto constantValueAttr = constantOp.getValue();
+  auto constantType = constantOp.getType();
+  if (constantValueAttr.isa<SplatElementsAttr>()) {
+    // Splats are always small and can often have special handling when we
+    // know they are a splat - which is why it's so important we inline them
+    // here so we know when they are used that's the case.
+    return true;
+  } else if (auto denseAttr = constantValueAttr.dyn_cast<DenseElementsAttr>()) {
+    // Smallish constants are worth moving inside.
+    auto shapedType = constantType.cast<ShapedType>();
+    uint64_t estimatedByteLength =
+        (shapedType.getNumElements() * shapedType.getElementTypeBitWidth()) / 8;
+    return denseAttr.isSplat() ||
+           estimatedByteLength <= kMaxInlinedConstantBytes;
+  } else if (constantType.isIntOrIndexOrFloat()) {
+    // Primitives can always go in.
+    return true;
+  }
+
+  return false;
+}
+
+// Returns true if the given value should be inlined into the closure region.
+// This is non-recursive and only holds for this value. Recursively cloning
+// trees is hard and it'd be better to model that differently such as by having
+// a wrapper region for immutable blobs that can be inlined that this then
+// returns true for.
+static bool shouldInlineIntoClosure(Value value) {
```

Comment:
Nice! I can add more things here (like `linalg.init_tensor`). Also does this subsume `RematerializeConstants...` Pass?